src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -101,6 +102,21 @@ internal MethodDesc MainMethod
             }
         }
 
+        private MethodDesc _startupCodeMainMethod;
+
+        internal MethodDesc StartupCodeMainMethod
+        {
+            get
+            {
+                if (_startupCodeMainMethod == null)
```

Comment:
I think we should create this in AddCompilationRootsForMainMethod, and not have the lazy initializer check here.